machine/v1beta1/types_machine.go

```diff
@@ -194,6 +199,38 @@ type MachineSpec struct {
 	ProviderID *string `json:"providerID,omitempty"`
 }
 
+// LifecycleHooks allow users to pause operations on the machine at
+// certain prefedined points within the machine lifecycle.
+type LifecycleHooks struct {
+	// PreDrain hooks prevent the machine from being drained.
+	// This also blocks further lifecycle events, such as termination.
+	// +optional
+	PreDrain []LifecycleHook `json:"preDrain,omitempty"`
+
+	// PreTerminate hooks prevent the machine from being terminated.
+	// PreTerminate hooks be actioned after the Machine has been drained.
+	// +optional
+	PreTerminate []LifecycleHook `json:"preTerminate,omitempty"`
+}
+
+// LifecycleHook represents a single instance of a lifecycle hook
+type LifecycleHook struct {
+	// Name defines a unique name for the lifcycle hook.
+	// The name should be unique and descriptive, ideally 1-3 words, in camelCase with no spaces.
+	// Names must be unique and should only be managed by a single entity.
+	// +kubebuilder:validation:Pattern:="[A-Za-z]+"
```

Comment:
no numbers, no slashes, no dots is intentional?  I'm ok with it if that is your preference.  Finalizers were envisioned as structured as, "foo.domain.name/item".

Reply:
Since the name becomes part of the Condtion Type, I had originally limited it so that it would just be standard alpha CamelCase as this is typically what I've seen in Condition Types. I'm not against extending it to be more but then it might conflict with the Name forming part of the Condition Type and having those fit to conventions

Reply:
From `Condition` in meta/v1

```
	// type of condition in CamelCase or in foo.example.com/CamelCase.
	// ---
	// Many .condition.type values are consistent across resources like Available, but because arbitrary conditions can be
	// useful (see .node.status.conditions), the ability to deconflict is important.
	// The regex it matches is (dns1123SubdomainFmt/)?(qualifiedNameFmt)
	// +required
	// +kubebuilder:validation:Required
	// +kubebuilder:validation:Pattern=`^([a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*/)?(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])$`
	// +kubebuilder:validation:MaxLength=316
	Type string `json:"type" protobuf:"bytes,1,opt,name=type"`

```

if you want to line it up to match.